walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.*;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    Collection<String> attendees = request.getAttendees();
+    ArrayList<TimeRange> invalidTimes = new ArrayList<>();
+    int fullDay = TimeRange.WHOLE_DAY.duration();
+
+    if(attendees.size() == 0) attendees = request.getOptionalAttendees();   
+    if (request.getDuration() > fullDay) return Arrays.asList();
+
+
+    invalidTimes = generateInvalidTimes(events,attendees);
+    invalidTimes = resolveOverlaps(invalidTimes);
+
+    ArrayList<TimeRange> validTimes = generateValidTimes(invalidTimes, request.getDuration());
+    ArrayList<TimeRange> invalidOptionalTimes = generateInvalidTimes(events, request.getOptionalAttendees());
+
+    
+    for(TimeRange invalid: invalidOptionalTimes) {
+      for(TimeRange valid : validTimes) {
+        if(invalid.equals(valid)) validTimes.remove(invalid);
+      }
+    }
+    TimeRange[] returnValidTimes = new TimeRange[validTimes.size()];
+    returnValidTimes = validTimes.toArray(returnValidTimes);
+
+    return Arrays.asList(returnValidTimes);
+  }
+
+  public ArrayList<TimeRange> resolveOverlaps(ArrayList<TimeRange> ranges) {
+    ListIterator<TimeRange> iterator = ranges.listIterator();
+
+    if(iterator.hasNext()){
+      TimeRange rangeOne = iterator.next();
+
+      if(iterator.hasNext()){
+        TimeRange rangeTwo = iterator.next();
+
+        if(rangeOne.equals(rangeTwo)){}
+
+        else{
+          if(rangeOne.overlaps(rangeTwo)){
+            int start = Math.min(rangeOne.start(), rangeTwo.start());
+            int end = Math.max(rangeOne.end(), rangeTwo.end());
+            int replaceIndex = ranges.indexOf(rangeOne);
+
+            TimeRange combinedRange = TimeRange.fromStartEnd(start, end, false);
+
+            ranges.set(replaceIndex, combinedRange);
+            ranges.remove(rangeTwo);
+          }
+        }
+      }
+    }
+    return ranges;
+  }
+
+  public ArrayList<TimeRange> generateValidTimes(ArrayList<TimeRange> ranges, long duration) {
```

Comment:
This method for finding free time based on empty time is very effective, but it's making an assumption: that the ranges are pre-sorted. This happens to be true in all the tests that were shipped with the task, but the following test will fail (I took an existing test, switched events around and then switched attendees around because you have a convenient sort on attendee that was making the test pass the first time I tried it):

```
@Test
  public void everyAttendeeIsConsideredMeetingsUnsorted() {
    // Have each person have different events. We should see two options because each person has
    // split the restricted times.
    //
    // Events  :       |--A--|     |--B--|
    // Day     : |-----------------------------|
    // Options : |--1--|     |--2--|     |--3--|

    Collection<Event> events = Arrays.asList(
        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
            Arrays.asList(PERSON_A)),
        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
          Arrays.asList(PERSON_B))
    );

    MeetingRequest request =
        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);

    Collection<TimeRange> actual = query.query(events, request);
    Collection<TimeRange> expected =
        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));

    Assert.assertEquals(expected, actual);
  }
```

Reply:
Wellll I guess I'll just have to make sure they're always sorted hehe. now I know why the "order_by_start/end" functions exists

Reply:
this should be fixed now